org.opent2t.sample.hub.superpopular/com.hue.hub/js/thingTranslator.js

```diff
@@ -19,39 +19,179 @@ class Translator {
+            headers: {
+                'cache-control': 'no-cache'
+            },
+            followAllRedirects: true,
+        };
+
+        return request(options)
+            .then(() => {
+                //Do Nothing
+            })
+            .catch((err) => {
+                if (err.statusCode == '401') {
+                    //extract nonce code from header
+                    var digestHeader = err.response.headers['www-authenticate'];
+                    var nonce = digestHeader.substr(digestHeader.indexOf('nonce=\"') + 7, 32);
+
+                    //send Digest Autehticaiton
+                    return this._sendDigestAuthentication(nonce, authInfo).then((body) => {
+                            var tokenInfo = JSON.parse(body); // This includes refresh token, scope etc..
+                            return new accessTokenInfo(
+                                        tokenInfo.access_token,
+                                        this._add2CurrentUTC(tokenInfo.access_token_expires_in),
+                                        tokenInfo.refresh_token,
+                                        this._add2CurrentUTC(tokenInfo.refresh_token_expires_in),
+                                        tokenInfo.token_type,
+                                        this._accessToken.bridgeId,
+                                        this._accessToken.whitelistId
+                                    );
+                        }).catch(function (err) {
+                            console.log("Request failed to: " + options.method + " - " + options.url);
+                            console.log("Error            : " + err.statusCode + " - " + err.response.statusMessage);
+                            throw err;
+                        });
+                } else {
+                    console.log("Request failed to: " + options.method + " - " + options.url);
+                    console.log("Error            : " + err.statusCode + " - " + err.response.statusMessage);
+                    throw err;
                 }
-            }
+            });
+     }
+
+     /**
+     * A helper function to the refreshAuthToken that send digest authentication to refresh token.
+     */
+    _sendDigestAuthentication(nonce, authInfo){
+                
+        //Compute digest header response
+        var HASH1 = crypto.createHash('md5').update(authInfo[0].client_id + ':oauth2_client@api.meethue.com:' + authInfo[0].client_secret).digest('hex');
+        var HASH2 = crypto.createHash('md5').update('POST:/oauth2/refresh').digest('hex');
+        var authHeaderResponse = crypto.createHash('md5').update(HASH1 + ':' + nonce + ':' + HASH2).digest('hex');
+        var disgestHeaderContent = 'username=\"' + authInfo[0].client_id 
+                                 + '\", realm=\"oauth2_client@api.meethue.com\", nonce=\"' + nonce 
+                                 + '\", uri=\"/oauth2/refresh\", response=\"' + authHeaderResponse + '\"';
```

Comment:
Sorry I don't quite get your comment here. What is the "this" part you are referring to?

---
In reply to: [94658455](https://github.com/openT2T/translators/pull/143#discussion_r94658455) [](ancestors = 94658455)